### DISCUSSION TITLE
I need help with adding a button to the main menu

### DISCUSSION THREAD

**Discussion by JonasunderscoreJones** [FabricMC/fabric-api#1795] [Mod Dev Support] [answered]
I'm trying to put a button on the title screen but i get an error message back claiming that the string can't be converted to text (the text I'm trying to display inside the button):

C:\Users\Jones\Desktop\mystery-mod-1.17.1\src\main\java\net\mystery\mod\mixin\TitleScreenMixin.java:23: error: incompatible types: String cannot be converted to Text        this.addButton(new ButtonWidget(this.width / 2 - 100 + 205, y, 50, 20, "bruh", (buttonWidget) -> {
This is my code:

package net.mystery.mod.mixin;

    import net.minecraft.client.MinecraftClient;
    import net.minecraft.client.gui.screen.Screen;
    import net.minecraft.client.gui.screen.TitleScreen;
    import net.minecraft.client.gui.screen.world.SelectWorldScreen;
    import net.minecraft.client.gui.widget.ButtonWidget;
    import net.minecraft.text.Text;
    import org.spongepowered.asm.mixin.Mixin;
    import org.spongepowered.asm.mixin.injection.At;
    import org.spongepowered.asm.mixin.injection.Inject;
    import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
    
    @Mixin(TitleScreen.class)
    public abstract class TitleScreenMixin extends Screen {
    
        protected TitleScreenMixin(Text title) {
            super(title);
        }
    
        @Inject(at = @At("RETURN"), method = "initWidgetsNormal")
        private void addCustomButton(int y, int spacingY, CallbackInfo ci) {
            this.addButton(new ButtonWidget(this.width / 2 - 100 + 205, y, 50, 20, "bruh", (buttonWidget) -> {
                MinecraftClient.getInstance().openScreen(new SelectWorldScreen(this));
            }));
        }
    }
I used [this](https://www.youtube.com/watch?v=SciS9-99-ZI) tutorial and I my code looks exactly the same (apart from the value of the string) but it doesn't work for me. Please help!

**Comment by Juuxel** [comment]
You need to use a Minecraft `Text` instead of a string (the tutorial is likely outdated in this case). A text can be one of
- `TranslatableText` (the preferred type of text since it works with multiple languages), which takes a translation key in a lang file
- `LiteralText`, which is just your string wrapped in a `Text` instance

(there are other types of `Text` too, but those are mostly irrelevant here).

Also, you can use Fabric API (fabric-screen-api-v1 to be specific) to add buttons to screens without mixins. [See the `ScreenEvents` javadoc for an example](https://github.com/FabricMC/fabric/blob/81e15a89363d705a2a5a7c1d8673348a2a7e1dfa/fabric-screen-api-v1/src/main/java/net/fabricmc/fabric/api/client/screen/v1/ScreenEvents.java#L86).

**Reply by JonasunderscoreJones** [reply]
It works now, thanks! I just used LiteralText but once the mod is complete I will probably change it to TranslatebleText. I will have a look into the fabric screen api but I think for the moment it might be a little to advanced for me as I have just startet modding and my java knowledge is very limited too

**Reply by tanchevk** [reply]
Just as a side-note: `TranslatableText` accepts any string as input, not just translation keys

**Reply by JFronny** [reply]
TranslatableText only accepts translation keys. If no translation is found, it will fall back to displaying the key. You can, of course, abuse this behavior by using the string you wish to be displayed as a translation key without a translation, but LiteralText is the correct class to use for that

**Comment by unatried** [comment, marked as answer]
Do this:
```java
package pack1.pack2.pack3;

import net.minecraft.client.gui.screen.Screen;
import net.minecraft.client.gui.screen.TitleScreen;
import net.minecraft.client.gui.screen.world.SelectWorldScreen;
import net.minecraft.client.gui.widget.ButtonWidget;
import net.minecraft.text.Text;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

@Mixin(TitleScreen.class)
public class HomeScreen extends Screen {


    protected HomeScreen(Text title) {
        super(title);
    }

    @Inject(at = @At("RETURN"), method = "addNormalWidgets")
    private void addModsButton(int y, int spacingY, CallbackInfoReturnable<Integer> cir) {
        this.addDrawableChild(ButtonWidget.builder(Text.translatable("modname.customButton"), (button) -> {
            this.client.setScreen(new SelectWorldScreen(this));
        }).dimensions(this.width / 2 - 200 +252, y, 50, 200).build());
    }
}
```
Inside of your resources folder (there where you have assets, the mixin file and the fabric.mod.json file) add the file: `./assets/modname/lang/en_US.json`
the content of the file:
```json
{
    "modname.customButton": "ButtonText"
}
```

**Comment by Tarkan-Zarrouk** [comment]
When adding a custom button in Minecraft 1.21.11 (Not sure if this works on upper versions) I found the following to work:

Inject at `createNormalMenuOptions`
render via `addRenderableWidget` using `Button.builder(...)`
Any text rendered in Minecraft is done here uses `Component.translatable(String)`
Then render your object via` setScreen`
On my part, `dimensions` was not found but `.bounds `was a viable solution for me!

<img width="1916" height="1080" alt="image" src="https://github.com/user-attachments/assets/3375f323-2c67-4935-8267-914137043406" />


Perhaps anyone working on 1.21.11 will find this useful!

**Comment by modmuss50** [comment]
You do not need a mixin to do this, you can use the `ScreenEvents.AFTER_INIT` event, see:

https://github.com/FabricMC/fabric-api/blob/867b876219518df0862da04ad8b21ed158799a70/fabric-screen-api-v1/src/testmodClient/java/net/fabricmc/fabric/test/screen/ScreenTests.java#L51C3-L67

**Reply by Tarkan-Zarrouk** [reply]
Ahhh! Much appreciated.

**Comment by hosseinpastilll-a11y** [comment]
Hi i have Dialog Window and Custommenu Plugin i want add the Buttom in the ESC Menu when i'am in my server my server is Paper 1.21.11 can someone help me?

**Reply by unatried** [reply]
This is the fabric discussions page not the paper discussions page.
